spec/ruby/optional/capi/ext/module_spec.c

```diff
@@ -71,8 +71,8 @@ static VALUE module_specs_rb_define_global_function(VALUE self, VALUE str_name)
   return Qnil;
 }
 
-static VALUE module_specs_rb_define_method(VALUE self, VALUE cls, VALUE str_name) {
-  rb_define_method(cls, RSTRING_PTR(str_name), module_specs_test_method, 0);
+static VALUE module_specs_rb_define_method(VALUE self, VALUE cls, VALUE str_name, VALUE argc) {
+  rb_define_method(cls, RSTRING_PTR(str_name), module_specs_test_method, FIX2INT(argc));
```

Comment:
Unfortunately one cannot just reuse a single function like this for different values of `argc` for `rb_define_method`.
The issue is that in C++ mode, the function pointer passed (`module_specs_test_method` here) must match the given argc. And also the last argument to `rb_define_method` must be static and not dynamic.

So we need something like
```c
VALUE module_specs_test_method2(VALUE self, VALUE arg1, VALUE arg2) {
  ...
}

static VALUE module_specs_rb_define_method2(VALUE self, VALUE cls, VALUE str_name) {
  rb_define_method(cls, RSTRING_PTR(str_name), module_specs_test_method2, 2);
}
```
Same for `-2` and `-1`, and those have different C arguments, see @tomstuart's comment and link.

You can see the errors with (on CRuby):
```
$ SPEC_CAPI_CXX=true mspec spec/ruby/optional/capi/module_spec.rb

module_spec.cpp: In function ‘VALUE module_specs_rb_define_method(VALUE, VALUE, VALUE, VALUE)’:
/home/eregon/.rubies/ruby-2.7.2/include/ruby-2.7.0/ruby/ruby.h:693:39: error: call to non-‘constexpr’ function ‘long int rb_fix2int(VALUE)’
  693 | #define RB_FIX2INT(x) ((int)rb_fix2int((VALUE)(x)))
      |                             ~~~~~~~~~~^~~~~~~~~~~~
/home/eregon/.rubies/ruby-2.7.2/include/ruby-2.7.0/ruby/ruby.h:2777:60: note: in definition of macro ‘rb_define_method’
 2777 | #define rb_define_method(m, n, f, a) rb_define_method_tmpl<a>::define(m, n, f)
      |                                                            ^
/home/eregon/.rubies/ruby-2.7.2/include/ruby-2.7.0/ruby/ruby.h:717:21: note: in expansion of macro ‘RB_FIX2INT’
  717 | #define FIX2INT(x)  RB_FIX2INT(x)
      |                     ^~~~~~~~~~
module_spec.cpp:75:74: note: in expansion of macro ‘FIX2INT’
   75 |   rb_define_method(cls, RSTRING_PTR(str_name), module_specs_test_method, FIX2INT(argc));
      |                                                                          ^~~~~~~
In file included from /home/eregon/.rubies/ruby-2.7.2/include/ruby-2.7.0/ruby.h:33,
                 from module_spec.cpp:1:
/home/eregon/.rubies/ruby-2.7.2/include/ruby-2.7.0/ruby/ruby.h:692:6: note: ‘long int rb_fix2int(VALUE)’ declared here
  692 | long rb_fix2int(VALUE);
      |      ^~~~~~~~~~
...
```